### PR TITLE
refactor(codegen): extract type-name utilities into ry::util and document compiler boundaries (#1820)

### PR DESCRIPTION
**Pull request by t0k0sh1**
## Summary

- Move 10 pure type-name helpers (`splitGenericTypeName`, `trimTypeNameSpaces`, `isList/Map/Set/Weak/Function/LowLevelTypeName`, `deriveRuntimeFnName`, `nativeSigKey`) from `CodeGen` static members into a new layer-independent `ry::util` namespace (`include/ry/util/type_name.hpp` + `src/util/type_name.cpp`). Mechanical move with no behavior change; all ~286 call sites across 21 files updated.
- Add `docs/architecture/{compiler-layers, llvm-ir-emission-boundary, runtime-abi-boundary}.md` to document the layer dependency direction, the candidate LLVM IR emission shared-library boundary (carry-over surface for #1949 / #1950), and the runtime ABI categorization for staged Rust migration.

Closes #1820

## Test plan

- [x] `cmake --build build` succeeds
- [x] `./build/ry_tests` (2523 pass / 0 fail; includes new `UtilTypeName.*` suite)
- [x] `./build/ry test -p` (203 Ry self-test files, 0 failures)
- [x] ASan+UBSan build + tests clean (0 sanitizer findings)
- [x] `grep -rn "CodeGen::splitGenericTypeName|..." src/ include/ tests/` returns empty
- [x] No double-prefix `ry::util::ry::util::` introduced

## Summary by CodeRabbit

* **Documentation**
  * Added comprehensive architecture docs for contributors covering compiler layer boundaries, LLVM IR emission interface, and runtime ABI.

* **Refactor**
  * Consolidated type-name parsing/utility logic into a shared utility layer and wired it across the codebase (no behavioral changes).

* **Tests**
  * Added a test suite validating the type-name utilities.

* **Chores**
  * Build updated to compile and run the new utilities and tests.

[Review Change Stack](https://app.coderabbit.ai/change-stack/t0k0sh1/ry/pull/1959?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: docs/architecture/compiler-layers.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+This document records the intended dependency direction between the compiler/runtime layers of the Ry implementation. It is the reference for issue #1820 (v0.0.26 boundary-tightening) and the foundation for the Rust migration path tracked in #1949 (LLVM IR emission shared library) and #1950 (Rust reimplementation).
+
+## Layers
+
+Source code is organized into the following layers, ordered from input (left) to output (right):
+
+```
+lexer → parser → AST → module loader → sema → codegen → runtime ABI
+```
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply docs/architecture/compiler-layers.md]: Applied: added `text` language tag to the fenced block at line 9 to satisfy markdownlint MD040.

File: src/util/type_name.cpp [hunk 2]
```diff
@@ -0,0 +1,103 @@
+#include "ry/util/type_name.hpp"
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply src/util/type_name.cpp:1]: Declining: the `codegen_*.cpp` / `runtime_*.cpp` / `parser_*.cpp` / `codegen_call_*.cpp` naming convention applies to the flat top-level `src/*.cpp` files only. `src/util/type_name.cpp` follows the existing subdirectory pattern shared by `src/cli/`, `src/jit/`, `src/parser/`, `src/sema/`, `src/source_manager/`, `src/runtime/{core,native}/`, `src/ast/`, etc. The new `src/util/` directory is the intentional layer-independent utility subdirectory introduced by #1820, paralleling the `src/runtime/{core,native}/` reorganization from #1829. See `docs/architecture/compiler-layers.md` "Invariants" section for the design rationale (layer-independent helpers live under `include/ry/util/` / `src/util/`).

File: src/util/type_name.cpp [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+                          std::string &head,
+                          std::vector<std::string> &inner) {
+    std::string t = trimWs(s);
+    size_t lt = t.find('<');
+    if (lt == std::string::npos) return false;
+    if (t.back() != '>') return false;
+    head = trimWs(t.substr(0, lt));
+    std::string body = t.substr(lt + 1, t.size() - lt - 2);
+    inner = splitTopLevelCommas(body);
+    return true;
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply src/util/type_name.cpp:72]: Declining for this PR: #1820 is a mechanical extraction of pure helpers from `CodeGen` into `ry::util` with no behavior change (per the approved plan). The pre-existing `CodeGen::splitGenericTypeName` returned `true` for `"<int>"`, `"List<>"`, and similar shapes; tightening that behavior would require auditing all ~286 call sites to confirm none depends on the lenient behavior, and that audit belongs in a focused follow-up issue rather than a stage-2 cleanup. The suggestion is good — happy to file it as a separate issue if a maintainer agrees the validation should be tightened.

File: src/util/type_name.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply src/util/type_name.cpp:1]: Thanks for confirming — the subdirectory carve-out is documented in `docs/architecture/compiler-layers.md` ("Layer-independent helpers" invariant) so future reviewers and contributors have a reference.

File: src/util/type_name.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply src/util/type_name.cpp:72]: Thanks for offering. Per our issue-filing policy, only the maintainer can authorize new issue creation, so I will not have an issue auto-opened here. If `@t0k0sh1` wants the `splitGenericTypeName` validation tightening tracked, they can file a follow-up themselves; otherwise it can be revisited when the next caller-side audit happens.
